### PR TITLE
Convert logging mutex functions to pybind11

### PR DESCRIPTION
**Pull request by sloretz**
Part of #665

Converts these functions to pybind11

* rclpy_logging_configure
* rclpy_logging_fini
* rclpy_create_node

### CODE REVIEW

**Comment by sloretz** [comment]
CI (build: `--packages-up-to rclpy` test: `--packages-select rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14059/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8879/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11751/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14150/)

File: rclpy/src/rclpy/logging.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+#include <mutex>
+#include <stdexcept>
+
+#include "rclpy_common/exceptions.hpp"
+#include "rclpy_common/handle.h"
+
+#include "logging.hpp"
+
+extern "C"
+{
```
Comment (hidmic) [inline rclpy/src/rclpy/logging.cpp:33]: @sloretz meta: why the `extern "C"` block? AFAIK it only affects the function symbol, C and C++ are ABI compatible.

File: rclpy/src/rclpy/node.cpp [hunk 2]
```diff
@@ -368,4 +372,139 @@ get_node_parameters(py::object pyparameter_cls, py::capsule pynode)
...
+  options.arguments = arguments;
+  options.enable_rosout = enable_rosout;
+
+  {
+    rclpy::LoggingGuard scoped_logging_guard;
+    ret = rcl_node_init(node.get(), node_name, namespace_, context, &options);
+  }
+
+  if (RCL_RET_BAD_ALLOC == ret) {
+    throw std::bad_alloc();
```
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:481]: @sloretz consider doing `rcl_reset_error()` before throwing.

File: rclpy/src/rclpy/logging.cpp (hunk 1 above)
Comment (hidmic) [reply rclpy/src/rclpy/logging.cpp:33]: On a related note, we should probably ensure symbol locality (i.e. `static`, anonymous namespace).

**Review by hidmic — COMMENTED** [review]
Overall LGTM !

File: rclpy/src/rclpy/node.cpp (hunk 2 above)
Comment (sloretz) [reply rclpy/src/rclpy/node.cpp:483]: Reset the rcl error in 336ca4f5803cb5fe4e1c6a88e9ea12556d115a57

File: rclpy/src/rclpy/logging.cpp (hunk 1 above)
Comment (sloretz) [reply rclpy/src/rclpy/logging.cpp:33]: Oops, I mistakenly thought extern "C" changed the calling convention, but it does not. Removed `extern C` stuff and made appropriate functions `static` in 570b63d7d1751b61099fd705fb22133e05ccdf38

**Comment by sloretz** [comment]
CI with feedback addressed
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14068/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8888/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11760/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14160/)

**Review by hidmic — APPROVED** [review]

**Comment by sloretz** [comment]
Rebased, and running windows CI to hopefully get a line number number for the warning

Windows: [Build Status](https://ci.ros2.org/job/ci_windows/14174/)

**Comment by sloretz** [comment]
Windows CI with hopefully a fix for the warning: [Build Status](https://ci.ros2.org/job/ci_windows/14177/)

File: rclpy/src/rclpy/logging.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  int severity,
+  const char * name,
+  rcutils_time_point_value_t timestamp,
+  const char * format,
+  va_list * args)
+{
+  try {
+    rclpy::LoggingGuard scoped_logging_guard;
+    rcl_logging_multiple_output_handler(location, severity, name, timestamp, format, args);
+  } catch (std::exception & ex) {
```
Comment (hidmic) [inline rclpy/src/rclpy/logging.cpp:55]: @sloretz nit: could be made `const`?

**Review by hidmic — APPROVED** [review]
LGTM !

File: rclpy/src/rclpy/logging.cpp (hunk 3 above)
Comment (sloretz) [reply rclpy/src/rclpy/logging.cpp]: Made exception catching const in 6dee6f8

**Comment by sloretz** [comment]
Rebased, CI (build: `--packages-above-and-dependencies rclpy` test: `--packages-above rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14116/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8924/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11805/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14213/)

**Review by hidmic — APPROVED** [review]

**Comment by sloretz** [comment]
CMake warning is in cyclondedds, not this PR. Merging :tada:

**Pull request merged**